### PR TITLE
C#: Model assertions in the CFG

### PR DESCRIPTION
**Pull request by hvitved**
### Overview

This PR models assertions directly in the CFG, which means that an assertion
```csharp
Assert(e);
```
gets roughly the same CFG as
```csharp
if (!e)
    throw new AssertionException();
```

This allows for simplifications to the guards/nullness libraries, as assertions are now just normal guards.

### Implementation

For an assertion
```csharp
Assert(e);
...
```
we want a CFG that looks like
```
              e
            /  \
           /    \
        true    false	 
         /        \
        /          \
       |            |   
call to Assert    call to Assert
       |            |
      ...          exit
```
which means that we need two copies of the `call to Assert` node. This is implemented using our CFG splitting framework.

### Commits

Commit-by-commit review is encouraged.

The implementation of assertions in the CFG (f1d6f7cd0c128092b0314a491ce585a910e20aa4) revealed an (unrelated) problem in the analysis of guards in the context of CFG splitting. This is witnessed by [a new test](https://github.com/github/codeql/pull/4293/commits/17f0ac4b208991ddbf4f04a0fe52b135ee1df043#diff-30ab0a2814437def50566a6f430c2502R343) added in the first commit:
```csharp
void M28(bool b)
{
    string s = b ? null : "";
    if (s != null && !b)
        Console.WriteLine(s.Length); // null guarded [MISSING]
}
```
The old analysis would require _all_ SSA variables for `s` in `s != null` to match an SSA variable for `s` in `s.Length`. However, we actually only need to check those SSA variables that are read in _any_ of the copies of `s != null` that guard `s.Length` (that is, only the copy for `b=false`). This change is implemented in the last commit, 1a93090778b05bb8914ad7d1b8051b9a3070af42.

### CSharp-Differences

https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/499

I want to highlight one fixed false positive: the access to `usingBinderOpt._syntax.Statement` in https://github.com/dotnet/roslyn/blob/d2244950fab50c039b3cdfc65da1243f39c326b1/src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs#L159 is no longer flagged as a potential NPE. This is because the statement is guarded by both [`Debug.Assert(isUsingDeclaration || usingBinderOpt != null)`](https://github.com/dotnet/roslyn/blob/d2244950fab50c039b3cdfc65da1243f39c326b1/src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs#L92) and [`!isUsingDeclaration`](https://github.com/dotnet/roslyn/blob/d2244950fab50c039b3cdfc65da1243f39c326b1/src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs#L153), and now that `isUsingDeclaration || usingBinderOpt != null` is modelled as a conditional expression, we get CFG splitting on the Boolean value `isUsingDeclaration`.

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowElement.qll [hunk 1]
```diff
@@ -128,37 +128,42 @@ class ControlFlowElement extends ExprOrStmtParent, @control_flow_element {
...
    */
   pragma[nomagic]
-  private predicate immediatelyControlsBlockSplit(BasicBlock succ, ConditionalSuccessor s) {
-    exists(ConditionBlock cb | this.immediatelyControlsBlockSplit0(cb, succ, s) |
-      forall(BasicBlock pred, SuccessorType t |
-        this.immediatelyControlsBlockSplit1(cb, succ, s, pred, t)
-      |
-        this.immediatelyControlsBlockSplit2(cb, succ, s, pred, t)
-      )
+  private predicate immediatelyControlsBlockSplit(
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowElement.qll:131]: The comment on this predicate is not mentioning this new parameter.

**Review by tamasvajk — APPROVED** [review]
Added some minor comments.

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll [hunk 2]
```diff
@@ -98,6 +99,13 @@ class Completion extends TCompletion {
     cfe instanceof ThrowElement and
     this = TThrowCompletion(cfe.(ThrowElement).getThrownExceptionType())
     or
+    exists(AssertMethod m | assertion(cfe, m, _) |
+      this = TThrowCompletion(m.getExceptionClass())
+      or
+      not exists(m.getExceptionClass()) and
```
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll:105]: We do, e.g. https://github.com/github/codeql/blob/58f51899c931af7abb90ca80f19523d6136880aa/csharp/ql/src/semmle/code/csharp/commons/Assertions.qll#L202.

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/Splitting.qll [hunk 3]
```diff
@@ -385,6 +387,135 @@ module InitializerSplitting {
...
+
+  private class AssertionSplitKind extends SplitKind, TAssertionSplitKind {
+    override int getListOrder() { result = InitializerSplitting::getNextListOrder() }
+
+    override predicate isEnabled(ControlFlowElement cfe) { this.appliesTo(cfe) }
+
+    override string toString() { result = "Assertion" }
+  }
+
+  int getNextListOrder() { result = InitializerSplitting::getNextListOrder() + 1 }
```
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/internal/Splitting.qll:441]: Each split kind must have a unique integer ID; `getNextListOrder` is used to specify the ID for one kind based on the kind that is syntactically before in the file.

**Review by tamasvajk — APPROVED** [review]

**Pull request merged**